mtp_api/apps/notification/rules.py

```diff
@@ -114,6 +118,63 @@ def get_event_trigger(self, record):
         return getattr(record, self.kwargs['profile'])
 
 
+class CountingRule(BaseRule):
+    def __init__(self, *args, profile, count, limit, **kwargs):
+        """
+        Counts a field (`count`) on records of the same type on related `profile` over the past 4 weeks.
+        If count exceeds `limit`, rule is triggered.
+        """
+        super().__init__(*args, profile=profile, count=count, limit=limit, **kwargs)
+        self.description = self.description.format(limit=limit)
+
+    @cached_property
+    def shared_profile(self):
+        # these are catch-all profiles for records that cannot have their own unique ones
+        # so counting rules would not be interesting or would need different limits
+        # c.f. `update_security_profiles` command
+        if self.kwargs['profile'] == 'sender_profile':
+            try:
+                return SenderProfile.objects.get_anonymous_sender()
+            except SenderProfile.DoesNotExist:
+                pass
+        elif self.kwargs['profile'] == 'recipient_profile':
+            try:
+                return RecipientProfile.objects.get_cheque_recipient()
+            except RecipientProfile.DoesNotExist:
+                pass
+        return None
+
+    def triggered(self, record):
+        profile = self.get_event_trigger(record)
+        if not profile or profile == self.shared_profile:
+            return False
+
+        records_of_same_type = self.get_profile_records_of_same_type(profile, record)
+        field_to_count = self.kwargs['count']
+        count = records_of_same_type.values(field_to_count).distinct().count()
```

Comment:
If only this count was kept updated and stored somewhere...

Reply:
where do i click to report abuse

Reply:
"this guy keeps commenting on my PRs and being right about everything"